src/luaotfload-harf-plug.lua

```diff
@@ -0,0 +1,1000 @@
+        -- the other discretionary handling, otherwise the discretionary
+        -- contents do not interact with the surrounding (e.g. no ligatures or
+        -- kerning) as it should.
+        if current and getid(current) == kern_t and getsubtype(current) == fontkern_t then
+          setprev(current, nil)
+          setnext(current, nil)
+          setfield(node, "replace", current)
+          head, current = removenode(head, current)
+        end
+        local pre, post, rep = getdisc(node)
+        setdisc(node, process(pre, fontid, direction),
+                      process(post, fontid, direction),
+                      process(rep, fontid, direction))
+
+        head, current = insertafter(head, current, node)
+      end
+      node = getnext(node)
+      nodeindex = nodeindex + 1
+    end
+  end
+  while node ~= run.after do
+    local oldnode = node
+    head, node = removenode(head, node)
+    freenode(oldnode)
+  end
+
+  return head, node
+end
+
+local function shape_run(head, current, run)
+  if not run.skip then
+    -- Font loaded with our loader and an HarfBuzz face is present, do our
+    -- shaping.
+    local fontid = run.font
+    local fontdata = font.getfont(fontid)
+    local options = fontdata.specification.features.raw
+
+    local glyphs
+    head, glyphs = shape(head, current, run)
+    return tonodes(head, current, run, glyphs)
+  else
+    return head, run.after
+  end
+end
+
+function process(head, font, direction)
+  local newhead, current = head, head
+  local runs = itemize(head, font, direction)
+
+  for _, run in next, runs do
+    newhead, current = shape_run(newhead, current, run)
+  end
+
+  return newhead or head
+end
+
+local function pdfdirect(data)
+  local n = newnode(whatsit_t, pdfliteral_t)
+  setfield(n, "mode", 2) -- direct
+  setdata(n, data)
+  return n
+end
+
+local function post_process(head)
+  for n in traverse(head) do
+    local props = getproperty(n)
+    local harfprops = props and props.harf
+
+    local startactual, endactual
+    if harfprops then
+      startactual = harfprops[startactual_p]
+      endactual = harfprops[endactual_p]
+    end
+
+    if startactual then
+      local actualtext = "/Span<</ActualText<FEFF"..startactual..">>>BDC"
+      head = insertbefore(head, n, pdfdirect(actualtext))
+    end
+
+    if endactual then
+      head = insertafter(head, n, pdfdirect("EMC"))
+    end
+
+    local replace = getfield(n, "replace")
+    if replace then
+      setfield(n, "replace", post_process(replace))
+    end
+
+    local subhead = getfield(n, "head")
+    if subhead then
+      setfield(n, "head", post_process(subhead))
+    end
+  end
+  return head
+end
+
+local function post_process_nodes(head, groupcode)
+  return tonode(post_process(todirect(head)))
+end
+
+local function run_cleanup()
+  -- Remove temporary PNG files that we created, if any.
+  -- FIXME: It would be nice if we wouldn't need this
+  for _, path in next, pngcachefiles do
+    osremove(path)
+  end
```

Comment:
luaotfload could cache the PNGs inside its font cache directories, but it would then need to handle the case when the font changes.